modules/cudastereo/include/opencv2/cudastereo.hpp

```diff
@@ -291,6 +291,97 @@ CV_EXPORTS Ptr<cuda::DisparityBilateralFilter>
+
+\f[C(d)\f]
+
+(This is implicitly summed over the whole image.)
+The problem is that many different depths will produce reasonable colors when pixels are considered individually so we add a term to the cost penalizing the difference between pixels and add a weighting factor:
+
+\f[f\mathbf{(\nabla\mathbf{d}) + \lambda C(d)}\f]
+
+where \f$\mathbf{\nabla d}\f$ is the difference between neighbouring pixels.
+
+@note
+   This is a slight abuse of notation since there are actually four neighbors for each pixel, so we evaluate the left term over each pair of neighbors and sum.
+
+One common choice for \f$f\f$ is the *Huber norm*, which is:
+
+\f[\left\| \nabla \mathbf{d} \right\|_e= \left\{\begin{matrix} \frac{(\nabla \mathbf{d})^2}{2\epsilon} & \mathrm{if} \left | \nabla \mathbf{d} \right | < \epsilon \\ \left |\nabla \mathbf{d} \right | - \frac{\epsilon}{2} &\mathrm{else} \end{matrix}\right\}\f]
+
+This is the same as a metal wire: it starts out acting like a spring, but when stretched too much it deforms plastically.
+
+This gives:
+
+\f[\left\| \nabla \mathbf{d} \right\|_e  + \lambda \mathbf{C(\mathbf{d})}\f]
+
+This problem is intractable to solve so we do a *relaxation*: we repeatedly solve the left and right sides independently, but enforce that the two solutions must be increasingly similar as we go along. We do this by creating a spring force between the two solutions. Remember from physics that a spring's energy is expressed as \f$\frac{1}{2} k(x_1-x_2)^2\f$, so we write:
+
+\f[\left\| \nabla \mathbf{d} \right\|_e + \frac{1}{2\theta} (\mathbf{d-a})^2  + \lambda \mathbf{C(\mathbf{a})}\f]
+
+
+\f$\frac{1}{\theta}\f$ is the spring constant, \f$\mathbf{d}\f$ is one solution, and \f$\mathbf{a}\f$ is the other. We refer to \f$\theta\f$ as the *stiffness*.
+
+We can also give a hint to the left hand side that certain places are likely to have discontinuities by varying the thickness of the wires:
+
+\f[\mathbf{g}\left\| \nabla \mathbf{d} \right\|_e + \frac{1}{2\theta} (\mathbf{d-a})^2  + \lambda \mathbf{C(\mathbf{a})}\f]
+
+The function \f$\mathbf{g}\f$ is the weight function.
+
+The right half is a literal search through all possible values of \f$\mathbf{C(\mathbf{a})}\f$ for each pixel.
+
+It turns out that solving the left half:
+
+\f[\mathbf{g}\left\| \nabla \mathbf{d} \right\|_e + \frac{1}{2\theta} (\mathbf{d-a}_{fixed})^2\f]
+
+is quite hard.
+
+*/
+class CV_EXPORTS DepthmapDenoiseWeightedHuber : public cv::Algorithm
+{
+public:
+    //! This may be called repeatedly to iteratively refine the internal depthmap
+    virtual cv::cuda::GpuMat operator()(InputArray input,
+                                        float epsilon,
+                                        float theta) = 0;
```

Comment:
Please rename this method to `refine` (or something similar). Since algorithm will be used via pointer, it will be more convenient to call `alg->refine(...)` instead of `(*alg)(...)` or `alg->operator()(...)`.